src/Symfony/Bridge/Doctrine/Form/Type/DoctrineType.php

```diff
@@ -165,22 +226,85 @@ public function configureOptions(OptionsResolver $resolver)
+                trigger_error('The "property" option is deprecated since version 2.7 and will be removed in 3.0. Use "choice_label" instead.', E_USER_DEPRECATED);
+            }
+
+            return $propertyName;
+        };
+
+        // Invoke the query builder closure so that we can cache choice lists
+        // for equal query builders
+        $queryBuilderNormalizer = function (Options $options, $queryBuilder) {
+            if (is_callable($queryBuilder)) {
+                $queryBuilder = call_user_func($queryBuilder, $options['em']->getRepository($options['class']));
+
+                if (!$queryBuilder instanceof QueryBuilder) {
+                    throw new UnexpectedTypeException($queryBuilder, 'Doctrine\ORM\QueryBuilder');
+                }
+            }
+
+            return $queryBuilder;
+        };
+
+        // deprecation note
+        $loaderNormalizer = function (Options $options, $loader) {
+            if ($loader) {
+                trigger_error('The "loader" option is deprecated since version 2.7 and will be removed in 3.0. Override getLoader() instead.', E_USER_DEPRECATED);
+            }
+
+            return $loader;
+        };
+
+        // Set the "id_reader" option via the normalizer. This option is not
+        // supposed to be set by the user.
+        $idReaderNormalizer = function (Options $options) use (&$idReaders) {
+            $hash = CachingFactoryDecorator::generateHash(array(
+                $options['em'],
+                $options['class'],
+            ));
+
+            // The ID reader is a utility that is needed to read the object IDs
+            // when generating the field values. The callback generating the
+            // field values has no access to the object manager or the class
+            // of the field, so we store that information in the reader.
+            // The reader is cached so that two choice lists for the same class
+            // (and hence with the same reader) can successfully be cached.
+            if (!isset($idReaders[$hash])) {
+                $classMetadata = $options['em']->getClassMetadata($options['class']);
+                $idReaders[$hash] = new IdReader($options['em'], $classMetadata);
+            }
+
+            return $idReaders[$hash];
+        };
+
         $resolver->setDefaults(array(
             'em' => null,
-            'property' => null,
+            'property' => null, // deprecated, use "choice_label"
             'query_builder' => null,
-            'loader' => $loader,
+            'loader' => null, // deprecated, use "choice_loader"
             'choices' => null,
-            'choice_list' => $choiceList,
-            'group_by' => null,
+            'choices_as_values' => true,
+            'choice_loader' => $choiceLoader,
+            'choice_label' => $choiceLabel,
+            'choice_name' => $choiceName,
+            'choice_value' => $choiceValue,
+            'id_reader' => null, // internal
```

Comment:
The problem is that we need to store the `IdReader` object (that depends on the "em" and "class" options) with the `Form` instance somehow. We only have two mechanisms for that:

* options
* attributes

Attributes don't work, since we need to access the ID reader already before the `Form` instance is built (from the other option closures). So this to me seems like the only alternative.